sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -618,6 +618,9 @@ object CombineLimits extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case ll @ Limit(le, nl @ Limit(ne, grandChild)) =>
       Limit(If(LessThan(ne, le), ne, le), grandChild)
+
+    case s @ Sort(_, _, Limit(exp, grandChild)) =>
```

Comment:
My bad, after thinking more about this, i think we can not add this rule, because push down sort leads to sort global data.
But here is another case we can optimize, i am updating this.